kvdb/sqlite/db.go

```diff
@@ -0,0 +1,207 @@
+)
+
+const (
+	// sqliteOptionPrefix is the string prefix sqlite uses to set various
+	// options. This is used in the following format:
+	//   * sqliteOptionPrefix || option_name = option_value.
+	sqliteOptionPrefix = "_pragma"
+
+	// sqliteTxLockImmediate is a dsn option used to ensure that write
+	// transactions are started immediately.
+	sqliteTxLockImmediate = "_txlock=immediate"
+)
+
+// NewSqliteBackend returns a db object initialized with the passed backend
+// config. If a sqlite connection cannot be established, then an error is
+// returned.
+func NewSqliteBackend(ctx context.Context, cfg *Config, dbPath, fileName,
+	prefix string) (walletdb.DB, error) {
+
+	// First, we add a set of mandatory pragma options to the query.
+	pragmaOptions := []struct {
+		name  string
+		value string
+	}{
+		{
+			name: "busy_timeout",
+			value: fmt.Sprintf(
+				"%d", cfg.BusyTimeout.Milliseconds(),
+			),
+		},
+		{
+			name:  "foreign_keys",
+			value: "on",
+		},
+		{
+			name:  "journal_mode",
+			value: "WAL",
+		},
+	}
+	sqliteOptions := make(url.Values)
+	for _, option := range pragmaOptions {
+		sqliteOptions.Add(
+			sqliteOptionPrefix,
+			fmt.Sprintf("%v=%v", option.name, option.value),
+		)
+	}
+
+	// Then we add any user specified pragma options.
+	for _, option := range cfg.PragmaOptions {
+		pair := strings.Split(option, "=")
+		if len(pair) != 2 {
+			return nil, fmt.Errorf("invalid pragma option: %v. "+
+				"Expected the format: name=value", option)
+		}
+
+		sqliteOptions.Add(
+			sqliteOptionPrefix,
+			fmt.Sprintf("%v=%v", pair[0], pair[1]),
+		)
+	}
+
+	// Construct the DSN which is just the database file name, appended
+	// with the series of pragma options as a query URL string. For more
+	// details on the formatting here, see the modernc.org/sqlite docs:
+	// https://pkg.go.dev/modernc.org/sqlite#Driver.Open.
+	writeDSN := fmt.Sprintf(
+		"%v?%v&%v", filepath.Join(dbPath, fileName),
+		sqliteOptions.Encode(), sqliteTxLockImmediate,
+	)
+	writeCfg := &sqlbase.Config{
+		DriverName:      "sqlite",
+		Dsn:             writeDSN,
+		Timeout:         cfg.Timeout,
+		TableNamePrefix: prefix,
+	}
+
+	writeDB, err := sqlbase.NewSqlBackend(ctx, writeCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	readDSN := fmt.Sprintf(
+		"%v?%v", filepath.Join(dbPath, fileName),
+		sqliteOptions.Encode(),
+	)
+	readCfg := &sqlbase.Config{
+		DriverName:      "sqlite",
+		Dsn:             readDSN,
+		Timeout:         cfg.Timeout,
+		TableNamePrefix: prefix,
+	}
+
+	readDB, err := sqlbase.NewSqlBackend(ctx, readCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sqliteDB{
+		readDBConn:  readDB,
+		writeDBConn: writeDB,
+	}, nil
+}
+
+// sqliteDB contains two sqlite connections, one is optimised for reads as it
+// uses the default BEGIN DEFERRED directive for starting a db transaction and
+// the other is optimised for writes as it uses the BEGIN IMMEDIATE directive
+// for starting a db transaction.
```

Comment:
I was reading up on sqlite transactions on https://www.sqlite.org/lang_transaction.html and saw this:

>While a read transaction is active, any changes to the database that are implemented by separate database connections will not be seen by the database connection that started the read transaction. If database connection X is holding a read transaction, it is possible that some other database connection Y might change the content of the database while X's transaction is still open, however X will not be able to see those changes until after the transaction ends. While its read transaction is active, X will continue to see an historic snapshot of the database prior to the changes implemented by Y.

To what extent does this apply to your setup with two connections? It seems different from bbolt where you're never reading an historic snapshot afaik?

Reply:
Good question re how connections work under the hood. I wish both sqlite and bbolt used more formal terminology here such as "snapshot isolation", or "linearizability", etc. AFACIT, this snapshot behavior seems to be identical to how bolt functions: read transactions can happen concurrently, they operate based on a snapshot of the database (the B+Tree is cloned from leaf up to root), write transactions can happen while read transactions are active. 

> Bolt allows only one read-write transaction at a time but allows as many read-only transactions as you want at a time. Each transaction has a consistent view of the data as it existed when the transaction started.


https://github.com/boltdb/bolt#transactions

Reply:
I see, yes that sounds mostly identical. What I don't think happens in bbolt is that read and write transactions actually run concurrently? So in bbolt the write only starts after all the reads have finished. For sqlite, it does sound a bit as if they can run in parallel, with the read txes having their snapshot. In lnd, there are sometimes non-database things happening inside a db tx block. Not sure if this may lead to differences in behavior with bbolt vs sqlite, given the potentially parallel transaction execution?

Reply:
> That I don't think happens in bbolt is that read and write transactions actually run concurrently?

My understanding is that there can only be a single writer at a time (there's a write mutex in the main db struct). Read transactions grab the read side of a RWMutex (for the mmap), while write transactions grab a write mutex (distinct from the mmap mutex, which needs to be obtained to update the memory map to commit).  The read transaction then operates on a _snapshot_ of the DB at that time. Readers also don't block other readers. If a write transaction is created, then readers can still be created (they copy the B+Tre root with the existing pointers). All other write transactions are blocked until that write finishes. 

An example trace through [bolt DB transactions](https://github.com/etcd-io/bbolt/blob/ece0e3dec20c7b22073c26535ebf6be6bcfaa384/db.go#L691-L694):
  * For read transactions, [a RWMutex is held for the mmap](https://github.com/etcd-io/bbolt/blob/ece0e3dec20c7b22073c26535ebf6be6bcfaa384/db.go#L703-L706)
  * [Root bucket gets copied](https://github.com/etcd-io/bbolt/blob/master/tx.go#L49-L57), which effectively copies the B+Tree root along w/ the other pointers (for both)
  * For write transactions, the write [lock is obtained](https://github.com/etcd-io/bbolt/blob/ece0e3dec20c7b22073c26535ebf6be6bcfaa384/db.go#L748-L750)

So with the above flow, a writer can start (grab the write block) and then get a snapshot copy, while several other readers also start and obtain a snapshot copy.

> . For sqlite, it does sound a bit as if they can run in parallel, with the read txes having their snapshot.

So this is basically the goal of the WAL: readers can look at the WAL, then start a transaction and ensure they apply the changes in the WAL. Writers just append to the end of the WAL. There's an index that has an exclusive lock on it, which needs to be consulted so both sides know what the latest state of the WAL is. 

My interpretation of the "database connection" terminology in the `sqlite` docs is the object return from the `sqlite.Open` call in C. `sqlite` has a few diff flags that [can be passed to open](https://www.sqlite.org/c3ref/open.html). Our library defaults to [SQLITE_OPEN_FULLMUTEX](https://www.sqlite.org/c3ref/c_open_autoproxy.html).

Reply:
I think this is the clearest set of docs from `sqlite` about their isolation model: https://www.sqlite.org/isolation.html. The WAL docs explain the read/write interaction (section 2.2): https://www.sqlite.org/wal.html. 

> If the same database is being read and written using two different [database connections](https://www.sqlite.org/c3ref/sqlite3.html) (two different [sqlite3](https://www.sqlite.org/c3ref/sqlite3.html) objects returned by separate calls to [sqlite3_open()](https://www.sqlite.org/c3ref/open.html)) and the two database connections do not have a [shared cache](https://www.sqlite.org/sharedcache.html), then the reader is only able to see complete committed transactions from the writer. Partial changes by the writer that have not been committed are invisible to the reader. This is true regardless of whether the two database connections are in the same thread, in different threads of the same process, or in different processes. This is the usual and expected behavior for SQL database systems.

Reply:
Thanks for all these pointers. It indeed seems that the original concern that I had is ungrounded.